widget/tabcontainer_test.go

```diff
@@ -47,12 +51,77 @@ func TestTabContainer_SelectTabIndex(t *testing.T) {
 	assert.Equal(t, 1, tabs.CurrentTabIndex())
 }
 
-func TestTabContainer_ApplyTheme(t *testing.T) {
-	tabs := NewTabContainer(&TabItem{Text: "Test1", Content: NewLabel("Test1")})
-	underline := Renderer(tabs).(*tabContainerRenderer).line
-	barColor := underline.FillColor
+func TestTabContainerRenderer_ApplyTheme(t *testing.T) {
+	tabs := widget.NewTabContainer(widget.NewTabItem("Test1", widget.NewLabel("Test1")))
+	var underline *canvas.Rectangle
+	driver.WalkObjectTree(tabs, fyne.NewPos(0, 0), func(o fyne.CanvasObject, offset fyne.Position) bool {
+		if u, ok := o.(*canvas.Rectangle); ok {
+			underline = u
+			return true
+		}
+		return false
+	}, nil)
+
+	r := widget.Renderer(tabs)
+
+	fyne.CurrentApp().Settings().SetTheme(theme.DarkTheme())
+	r.ApplyTheme()
+	darkColor := underline.FillColor
 
 	fyne.CurrentApp().Settings().SetTheme(theme.LightTheme())
-	Renderer(tabs).ApplyTheme()
-	assert.NotEqual(t, barColor, underline.FillColor)
+	r.ApplyTheme()
+	lightColor := underline.FillColor
+
+	assert.NotEqual(t, darkColor, lightColor)
+}
+
+func TestTabContainerRenderer_Layout(t *testing.T) {
+	tabs := widget.NewTabContainer(
+		widget.NewTabItemWithIcon("Text1", theme.CancelIcon(), canvas.NewCircle(theme.BackgroundColor())),
+		widget.NewTabItemWithIcon("Text2", theme.ConfirmIcon(), canvas.NewCircle(theme.BackgroundColor())),
+	)
+
+	r := widget.Renderer(tabs)
+	r.Layout(fyne.NewSize(100, 100))
+
+	type dim struct {
+		fyne.Position
+		fyne.Size
+	}
+	images := []dim{}
+	texts := []dim{}
+	driver.WalkObjectTree(tabs, fyne.NewPos(0, 0), func(o fyne.CanvasObject, offset fyne.Position) bool {
+		switch o.(type) {
+		case *canvas.Image:
+			images = append(images, dim{Position: o.Position().Add(offset), Size: o.Size()})
+		case *canvas.Text:
+			texts = append(texts, dim{Position: o.Position().Add(offset), Size: o.Size()})
+		}
+		return false
+	}, nil)
+
+	if assert.Len(t, images, 2) && assert.Len(t, texts, 2) {
```

Comment:
Surely here we should have the inverse logic and fail if there were images or texts missing?
The the rest will execute if the test did not fail...

Reply:
I don't get the point. If there are images or texts missing, the length of the slices is not two and the assertions fail (and the test, too).
The same if there are extra texts/images.

Reply:
Apologies, I should have been more clear. I prefer to keep complexity (nested depth) down.
If you return if either of those assertions fail then you can return early. Then the remaining code happens at the top level of the function.
"exit early" is one of the mantras I like to follow.

Reply:
Exitiing early here will not change anything on the complexity. You still have an `if`.
But I could switch from `assert` to `require` and leave out the `if`.

Reply:
The eager return is a reasonably typical approach to keeping complexity down.
It's main aim is that the core code of a method should be at the top level. Here the if is determining if the next code can run. Therefore it is neater to exit if the code cannot continue and then have the majority of this code at it's top level.